### PR TITLE
Feat: Evaluation Environment Refactor

### PR DESCRIPTION
**Pull request by AbdelrahmanAbdelgwad**
## 📜 Summary
The need to change the render rate of the evaluation is just as important as the training to allow for more flexibility for the user. Also the evaluation environment required some adjustments to conform to the last major code refactor.

###### Ticket(s): [Add render-each arg to eval env](https://app.clickup.com/t/8669k5dv2)

- 🔄 **Main changes in this PR**
    - Added render-each-eval arg 
    - Overhauled the eval_env to conform to the last code refactor

## 📝 Notes for future improvements

- Allow for the customization of the evaluation by allowing for various evaluation configurations instead of just one.
- Add more evaluation scenarios

### CODE REVIEW

File: src/highrl/envs/eval_env.py [hunk 1]
```diff
@@ -3,98 +3,181 @@
...
+from highrl.utils.abstract import Position
 from highrl.envs.robot_env import RobotEnv
 from highrl.obstacle.single_obstacle import SingleObstacle
 
 
 class RobotEvalEnv(RobotEnv):
+    """Evaluation environment for robot performance"""
+
     def __init__(self, config: RawConfigParser, args: Namespace):
         super(RobotEvalEnv, self).__init__(config, args)
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:15]: Rewrite this to conform with the new PEP8 style rules:
```python
super().__init__(config, args)
```

File: src/highrl/envs/eval_env.py [hunk 2]
```diff
@@ -3,98 +3,181 @@
...
 
+from highrl.utils.abstract import Position
 from highrl.envs.robot_env import RobotEnv
 from highrl.obstacle.single_obstacle import SingleObstacle
 
 
 class RobotEvalEnv(RobotEnv):
+    """Evaluation environment for robot performance"""
+
     def __init__(self, config: RawConfigParser, args: Namespace):
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:14]: Add a return type for the `__init__`, i.e. `-> None`.

File: src/highrl/envs/eval_env.py [hunk 3]
```diff
@@ -3,98 +3,181 @@
...
-            self.num_of_hard_obstacles,
-            min_dim=200,
-            max_dim=400,
-        )
-        self._generate_obstacles_points(
-            self.num_of_medium_obstacles,
-            min_dim=150,
-            max_dim=300,
+        self.cfg.n_eval_episodes = config.getint("eval", "n_eval_episodes")
+        self.add_boarder_obstacles()
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:17]: Misspelling here in `self.add_boarder_obstacles()`. Change it to `self.add_border_obstacles()`. 

Moreover, shouldn't this function be called at the `RobotEnv.__init__` ?

File: src/highrl/envs/eval_env.py [hunk 4]
```diff
@@ -3,98 +3,181 @@
...
+        robot_init_x_pos = config.getint("eval", "robot_init_x_pos")
+        robot_init_y_pos = config.getint("eval", "robot_init_y_pos")
+        goal_x_pos = config.getint("eval", "goal_x_pos")
+        goal_y_pos = config.getint("eval", "goal_y_pos")
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:21]: Those shouldn't be here, they should be added to `RobotConfigs` in `highrl.utils.general.py`.

File: src/highrl/envs/eval_env.py [hunk 5]
```diff
@@ -3,98 +3,181 @@
...
+            Position[int](robot_init_x_pos, robot_init_y_pos),
+            Position[int](goal_x_pos, goal_y_pos),
         )
-        self._generate_obstacles_points(
-            self.num_of_small_obstacles,
-            min_dim=50,
-            max_dim=150,
+        big_obs_count = config.getint("eval", "big_obs_count")
+        med_obs_count = config.getint("eval", "med_obs_count")
+        sml_obs_count = config.getint("eval", "sml_obs_count")
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:28]: Same here, those be placed into `highrl.utils.general.py`.

File: src/highrl/envs/eval_env.py [hunk 6]
```diff
@@ -3,98 +3,181 @@
...
+
+        # big_obs_pos = config.get("eval", "big_obs_pos")
+        # med_obs_pos = config.get("eval", "med_obs_pos")
+        # sml_obs_pos = config.get("eval", "sml_obs_pos")
+
+        big_obs_pos = [(180, 180), (125, 125)]
+        med_obs_pos = [(125, 70), (70, 125), (200, 50), (150, 30)]
+        sml_obs_pos = [(30, 10), (10, 50), (100, 80), (20, 170)]
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:36]: It is NOT wise to hard-code those values, as these will change in the future when we incorporate multiple evaluation environments with different difficulties.

File: src/highrl/envs/eval_env.py [hunk 7]
```diff
@@ -3,98 +3,181 @@
...
+
+        big_obs_dim = config.getint("eval", "big_obs_dim")
+        med_obs_dim = config.getint("eval", "med_obs_dim")
+        sml_obs_dim = config.getint("eval", "sml_obs_dim")
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:40]: Same here, those should be place into `highrl.utils.general.py`.

File: src/highrl/envs/eval_env.py [hunk 8]
```diff
@@ -3,98 +3,181 @@
...
+            med_obs_dim,
+            sml_obs_dim,
         )
 
-    def _generate_obstacles_points(
-        self, obstacles_count: int, min_dim: int, max_dim: int
-    ) -> None:
-        """Generate obstacles based on teacher action for next robot session
+        for obstacle in obstacles:
+            self.obstacles.obstacles_list.append(obstacle)
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:55]: Maybe this logic should be added as a function in the `highrl.obstacles.Obstacles` class? 
```python
Class Obstacles:
...
   def add_obstacles(self, obstacles_list: List[SingleObstacle]) -> None:
       """Add new obstacles to the current obstacles list"""
       self.obstacles_list.extend(obstacles_list)
```
This way the implementation detail (i.e., the obstacles are store in a list) is hidden from the user.

File: src/highrl/envs/eval_env.py [hunk 9]
```diff
@@ -3,98 +3,181 @@
...
+        big_obs_count,
+        med_obs_count,
+        sml_obs_count,
+        big_obs_pos,
+        med_obs_pos,
+        sml_obs_pos,
+        big_obs_dim,
+        med_obs_dim,
+        sml_obs_dim,
+    ):
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:68]: Add type hints and a doc-string here, please.

File: src/highrl/envs/eval_env.py [hunk 10]
```diff
@@ -3,98 +3,181 @@
...
+#         self.lidar_angle_increment = config.getfloat("lidar", "lidar_angle_increment")
+#         self.lidar_min_angle = config.getfloat("lidar", "lidar_min_angle")
+#         self.lidar_max_angle = config.getfloat("lidar", "lidar_max_angle")
+
+#         self.collision_score = config.getint("reward", "collision_score")
+#         self.reached_goal_score = config.getint("reward", "reached_goal_score")
+#         self.minimum_velocity = config.getfloat("reward", "minimum_velocity")
+#         self.minimum_distance = config.getfloat("reward", "minimum_distance")
+#         self.maximum_distance = config.getfloat("reward", "maximum_distance")
+#         self.velocity_std = config.getfloat("reward", "velocity_std")
```
Comment (ahmedheakl) [inline src/highrl/envs/eval_env.py:131]: What is the point of all of these comments, you can just remove them!

File: src/highrl/envs/teacher_env.py [hunk 11]
```diff
@@ -8,7 +8,7 @@
 from gym import Env, spaces
 import numpy as np
 from prettytable import PrettyTable
-
+from highrl.utils.abstract import Position
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:11]: You can just do
```python
from highrl.utils import Position
```

File: src/highrl/utils/training_utils.py [hunk 12]
```diff
@@ -120,14 +120,23 @@ def start_robot_session(
...
     )
+    # eval_callback = robot_callback.RobotEvalCallback(
+    #     eval_env=opt.eval_env,
+    #     n_eval_episodes=cfg.n_robot_eval_episodes,
+    #     logpath=eval_logpath,
+    #     savepath=eval_model_save_path,
+    #     eval_frequency=cfg.max_session_timesteps,
+    #     verbose=1,
+    #     render=cfg.render_eval,
+    # )
```
Comment (ahmedheakl) [inline src/highrl/utils/training_utils.py:131]: There is no point of saving all of these comments!

File: src/highrl/utils/training_utils.py [hunk 13]
```diff
@@ -120,14 +120,23 @@ def start_robot_session(
...
     eval_callback = robot_callback.RobotEvalCallback(
         eval_env=opt.eval_env,
-        n_eval_episodes=cfg.n_robot_eval_episodes,
+        n_eval_episodes=opt.eval_env.cfg.n_eval_episodes,
         logpath=eval_logpath,
         savepath=eval_model_save_path,
         eval_frequency=cfg.max_session_timesteps,
         verbose=1,
-        render=cfg.render_eval,
+        render=1,
```
Comment (ahmedheakl) [inline src/highrl/utils/training_utils.py:139]: Shouldn't this be opt.eval_env.cfg.render_each_eval`?

**Review by ahmedheakl — CHANGES_REQUESTED** [review]
Please fix request changes.

File: src/highrl/envs/eval_env.py (hunk 1 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/eval_env.py:15]: Done

**Review by ahmedheakl — CHANGES_REQUESTED** [review]
LGTM!

**Review by ahmedheakl — APPROVED** [review]
LGTM!

**Pull request merged**